atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidatorTest.java

```diff
@@ -0,0 +1,320 @@
+            CassandraServer.of(InetSocketAddress.createUnresolved("1.1.1.1", 80));
+    private static final String NEW_HOST_ONE = "new_host_one";
+    private static final String NEW_HOST_TWO = "new_host_two";
+    private static final String NEW_HOST_THREE = "new_host_three";
+    private static final Set<String> NEW_HOSTS = ImmutableSet.of(NEW_HOST_ONE, NEW_HOST_TWO, NEW_HOST_THREE);
+
+    private static final String OLD_HOST_ONE = "old_host_one";
+    private static final String OLD_HOST_TWO = "old_host_two";
+    private static final String OLD_HOST_THREE = "old_host_three";
+    private static final Set<String> OLD_HOSTS = ImmutableSet.of(OLD_HOST_ONE, OLD_HOST_TWO, OLD_HOST_THREE);
+    private static final Set<String> UUIDS = ImmutableSet.of("uuid1", "uuid2", "uuid3");
+
+    private static final Set<String> ALL_HOSTS = Sets.union(NEW_HOSTS, OLD_HOSTS);
+
+    private static final HostIdResult DEFAULT_RESULT = HostIdResult.success(UUIDS);
+
+    private final CassandraTopologyValidationMetrics metrics =
+            CassandraTopologyValidationMetrics.of(new DefaultTaggedMetricRegistry());
+    private final CassandraTopologyValidator validator = spy(new CassandraTopologyValidator(metrics));
+
+    @Test
+    public void retriesUntilNoNewHostsReturned() {
+        Iterator<String> uuidIterator = UUIDS.iterator();
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(NEW_HOSTS);
+        EntryStream.of(allHosts)
+                .values()
+                .forEach(container -> setHostIds(
+                        Set.of(container),
+                        HostIdResult.success(Set.of(uuidIterator.next())),
+                        HostIdResult.success(UUIDS)));
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(), allHosts, Duration.ofMillis(1), Duration.ofSeconds(20)))
+                .isEmpty();
+    }
+
+    @Test
+    public void retriesMarksFailureWhenNeverResolves() {
+        Iterator<String> uuidIterator = UUIDS.iterator();
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(NEW_HOSTS);
+        EntryStream.of(allHosts)
+                .values()
+                .forEach(container -> setHostIds(Set.of(container), HostIdResult.success(Set.of(uuidIterator.next()))));
+        assertThat(validator.getNewHostsWithInconsistentTopologiesAndRetry(
+                        allHosts.keySet(), allHosts, Duration.ofMillis(1), Duration.ofMillis(1)))
+                .isNotEmpty();
+        assertThat(metrics.validationFailures().getCount()).isEqualTo(1);
+        assertThat(metrics.validationLatency().getCount()).isEqualTo(1);
+    }
+
+    @Test
+    public void retiresAndReturnsFailingHosts() {
+        Map<CassandraServer, CassandraClientPoolingContainer> allHosts = setupHosts(NEW_HOSTS);
+        doReturn(allHosts.keySet()).when(validator).getNewHostsWithInconsistentTopologies(any(), any());
```

Comment:
nit: might be missing something, does `when(validator.getNewHostsWithInconsistentTopologies(any(), any()).thenReturn(allHosts.keySet());` not work? This is more common (and you've also used it later in the PR)

Reply:
I agree with your preference, but it is my understanding that for spied classes you must use `doReturn` otherwise `when(validator.getNewHostsWithInconsistentTopologies(any(), any())` actually invokes the underlying method.